compiler-core/src/javascript/typescript.rs

```diff
@@ -0,0 +1,530 @@
+                Statement::Fn { .. }
+                | Statement::TypeAlias { .. }
+                | Statement::CustomType { .. }
+                | Statement::ExternalType { .. }
+                | Statement::ExternalFn { .. }
+                | Statement::ModuleConstant { .. } => (),
+
+                Statement::Import {
+                    module, package, ..
+                } => {
+                    self.register_import(&mut imports, package, module);
+                }
+            }
+        }
+
+        imports
+    }
+
+    /// Registers an import of an external module so that it can be added to
+    /// the top of the generated Document. The module names are prefixed with a
+    /// "$" symbol to prevent any clashes with other Gleam names that may be
+    /// used in this module.
+    ///
+    fn register_import(
+        &mut self,
+        imports: &mut Imports<'a>,
+        package: &'a str,
+        module: &'a [String],
+    ) {
+        let module_name = module
+            .last()
+            .expect("Typescript generator could not identify imported module name.");
+        let module_name = format!("${}", module_name);
+        let path = self.import_path(package, module);
+        imports.register_module(path, [module_name], []);
+    }
+
+    /// Calculates the path of where to import an external module from
+    ///
+    fn import_path(&self, package: &'a str, module: &'a [String]) -> String {
+        let path = module.join("/");
+
+        // TODO: strip shared prefixed between current module and imported
+        // module to avoid decending and climbing back out again
+        if package == self.module.type_info.package || package.is_empty() {
+            // Same package
+            match self.module.name.len() {
+                1 => format!("./{}.d.ts", path),
+                _ => {
+                    let prefix = "../".repeat(self.module.name.len() - 1);
+                    format!("{}{}.d.ts", prefix, path)
+                }
+            }
+        } else {
+            // Different package
+            let prefix = "../".repeat(self.module.name.len() + 1);
+            format!("{}{}/dist/{}.d.ts", prefix, package, path)
+        }
+    }
+
+    fn statement(&mut self, statement: &'a TypedStatement) -> Vec<Output<'a>> {
+        match statement {
+            Statement::TypeAlias {
+                alias,
+                public,
+                type_,
+                ..
+            } if *public => vec![self.type_alias(alias.clone(), type_)],
+            Statement::TypeAlias { .. } => vec![],
+
+            // TODO: Do we need to do anything here?
+            Statement::ExternalType { .. } => vec![],
+
+            Statement::Import { .. } => vec![],
+
+            Statement::CustomType {
+                public,
+                constructors,
+                opaque,
+                name,
+                ..
+            } if *public && !*opaque => self.custom_type_definition(name, constructors),
```

Comment:
We would want to define the type but not the constructors if it is opaque and public

Reply:
So the idea is that in the Typescript declaration file it would look something like this?

Input:
```gleam
pub opaque type Animal {
  Cat(name: String)
  Dog(name: String)
} 
```

Output:
```typescript
import * as $gleam from "../gleam.d.ts";

class Cat extends $gleam.CustomType {
  constructor(name: string);
  name: string;
}

class Dog extends $gleam.CustomType {
  constructor(name: string);
  name: string;
}

export type Animal$ = Cat | Dog;
```

So the type is "exposed" but no way to import the individual constructors?

Reply:
If that's valid typescript that would be great. Sorry for taking so long to reply, I missed this one